README.md
Particle Swarm Optimisation: A python implementation
============================================

Python code that is used to implement particle swarm optimisation (PSO). PSO is an optimisation algorithm based on the flocking patterns of birds. It is a simple optimisation scheme that does not require complex operations. This file will provide details on how to use set up and use the python PSO file. 


---

[Requirements](#requirements) | [Use](#use) | [Quantitative
features](#quantitative-features) | [Files](#files) | [Test computer
setup](#test-computer-setup) | [Licence](#licence) | [References](#references) |
[Contact](#contact)

## Requirements:
Python 3.7 or newer with the following packages installed
| Package       | Version installed  |
| ------------- |:------------------:|
| numpy         | 1.17.1             |


## Use 

Set path `(Particle_swarm_optimisation)` to the folder location of Particle_swarm_optimisation. Then the following will load the main functions:
```
  import sys
  sys.path.append(Particle_swarm_optimisation)
  import PSO
```

As an example, we will use PSO to find the centre of a circle of the form (x-a)^2 + (y-b)^2 = r^2 where x, y and r equal 3, 4 and 5 respectfully:
```
  import numpy as np

  # Define the cost function
  def circle_cost(positions, args):
    return np.abs(25 - ((3 - positions[0])**2 + (4 - positions[1])**2))
  cost_function = circle_cost

  # Set maximum values of a and b
  varmin = [-1000, -1000]  # a min, b min
  varmax = [1000, 1000]    # a max, b max

  # Create problem dictionary
  problem = {'cost_function': cost_function, 'varmin': varmin, 'varmax': varmax}

  # Create PSO parameters dictionary
  params = {'maxit': 15, 'n_part': 30, 'c1': 2, 'c2': 2, 'w': 0.9, 'wdamp': 0.2, 'con': 1, 'display_info': 1,
            'early_stopping': 1, 'early_stopping_rounds': 7}

  # Run PSO and return lowest cost, a,b values and the costs per iteration
  gbest_value, gbest_position, best_costs = PSO.PSO(problem, params)

```

The PSO function takes in 2 arguments and the possibility to add more in the form of *args if the cost function needs them.


The first parameter is the `problem` parameter, it is a dictionary and the following values are needed in the dictionary:
| Value            | Purpose                               |
| ---------------- |:-------------------------------------:|
| 'cost_function'  | Cost function that PSO is to evaluate |
| 'varmin'         | Minimum possible particle locations   |
| 'varmax'         | Maximum possible particle locations   |



The second parameter is the `params` parameter, it is a dictionary and has lots of default parameters so they do not need to be specified all the time:
| Value                   | Purpose                                          | Default  |
| ----------------------- |:------------------------------------------------:|:--------:|
| 'maxit'                 | Max number of PSO iterations                     |30        |
| 'n_part'                | Number of PSO particles                          |15        |
| 'w'                     | Inertia weight                                   |1         |
| 'wdamp'                 | Inertia weight damping factor                    |1         |
| 'c1'                    | Personal constant                                |2.05      |
| 'c2'                    | Social constant                                  |2.05      |
| 'con'                   | Constriction factor                              |0.7298    |
| 'display_info'          | Display information after each iteration         |1         |
| 'early_stopping'        | Use early stopping                               |0         |
| 'early_stopping_rounds' | How many rounds to stop after if no improvement  |10        |
| 'velocity_limit_scale'  | Scaling factor used to set max velocity          |0.2       |
| 'BPSO'                  | Is it binary PSO                                 |0         |





Note: care should be taken when using binary PSO. The `varmin` and `varmax` values in the `problem` dictionary should 
not be set to 0 and 1 but rather values like -5 and 5. The velocities are still used to assess how strong of a draw 
there is to being 1 or 0. It is also recommended to set 'velocity_limit_scale' to 1 for binary PSO.





## Files
Some python files (.py files) have a description and an example in the header. To read this
header, type `help(filename)` in the console after importing (`import filename`).  Directory structure is as follows: 
```
├── demos.py              # Python file with examples
├── CHANGELOG.md          # changelog file
├── LICENSE.md            # license file 
├── PSO.py                # PSO file
└── README.md             # readme file describing project
```




## Test computer setup
- hardware:  Intel Core i7-8700K @ 3.2GHz; 32GB memory.
- operating system: Windows 10 64-bit
- software: python 3.7


## Licence

```
Copyright (c) 2020, Brian M. Murphy, University College Cork
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

  Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

  Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

  Neither the name of the University College Cork nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```


## References

1. Kennedy, James, and Russell Eberhart. "Particle swarm optimization." Proceedings of ICNN'95-International Conference on Neural Networks. Vol. 4. IEEE, 1995.

2. Shi, Yuhui, and Russell Eberhart. "A modified particle swarm optimizer." 1998 IEEE international conference on evolutionary computation proceedings. IEEE world congress on computational intelligence (Cat. No. 98TH8360). IEEE, 1998.

3. Clerc, Maurice, and James Kennedy. "The particle swarm-explosion, stability, and convergence in a multidimensional complex space." IEEE transactions on Evolutionary Computation 6.1 (2002): 58-73.

4. Kennedy, James, and Russell C. Eberhart. "A discrete binary version of the particle swarm algorithm." 1997 IEEE International conference on systems, man, and cybernetics. Computational cybernetics and simulation. Vol. 5. IEEE, 1997.




## Contact

Brian M. Murphy

Neonatal Brain Research Group,  
[INFANT Research Centre](https://www.infantcentre.ie/),  
Department of Paediatrics and Child Health,  
Room 2.18 UCC Academic Paediatric Unit, Cork University Hospital,  
University College Cork,  
Ireland

- email: Brian.M.Murphy AT umail dot ucc dot ie 

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. Project first added to GitHub on
2020-02-20. 



## [0.2.0] - 2020-05-18
### Added
- Added option to use binary PSO
- Added a demo file that contains 1 example


## [0.1.0] - 2020-02-20
### Added
- Added working version of particle swarm optimisation
- Using [semantic versioning](http://semver.org/) 




[0.1.0]: https://github.com/BrianMur92/Particle_swarm_optimisation/releases/tag/0.1.0
[0.2.0]: https://github.com/BrianMur92/Particle_swarm_optimisation/releases/tag/0.2.0

demos.py
# Author:       Brian Murphy
# Date started: 07/05/2020
# Last updated: <06/05/2020 05:49:32 (BrianM)>

import numpy as np
import PSO


def find_the_centre_of_a_circle():
    """
    Find the centre of a circle of the form (x-a)^2 + (y-b)^2 = r^2 where x, y and r equal 3, 4 and 5 respectfully

    Example - how to run:
        import demos
        demos.find_the_centre_of_a_circle()

    """
    # Define the cost function
    def circle_cost(positions, args):
        return np.abs(25 - ((3 - positions[0]) ** 2 + (4 - positions[1]) ** 2))

    cost_function = circle_cost

    # Set maximum values of a and b
    varmin = [-1000, -1000]  # a min, b min
    varmax = [1000, 1000]  # a max, b max

    # Create problem dictionary
    problem = {'cost_function': cost_function, 'varmin': varmin, 'varmax': varmax}

    # Create PSO parameters dictionary
    params = {'maxit': 15, 'n_part': 30, 'c1': 2, 'c2': 2, 'w': 0.9, 'wdamp': 0.2, 'con': 1, 'display_info': 1,
              'early_stopping': 1, 'early_stopping_rounds': 7}

    # Run PSO and return lowest cost, a,b values and the costs per iteration
    gbest_value, gbest_position, best_costs = PSO.PSO(problem, params)

    print('Found the centre of the circle as a = %f and b = %f' % (gbest_position[0], gbest_position[1]))
    print('The error was %f' % gbest_value)



PSO.py
# Author:       Brian Murphy
# Date started: 16/03/2020
# Last updated: <08/05/2020 09:52:57 (BrianM)>

import numpy as np
import time


def default_PSO_params(params):
    """
    This is to set the default PSO parameters. Note by default constriction will be applied over inertia weighting.
    To prevent this, specify the wdamp parameter that isn't 1.

    Will display info by default

    Early stopping is disabled by default
    """
    if 'maxit' not in params:
        params['maxit'] = 30
    if 'n_part' not in params:
        params['n_part'] = 15
    if 'w' not in params:
        params['w'] = 1
    if 'wdamp' not in params:
        params['wdamp'] = 1
    if 'c1' not in params:
        params['c1'] = 2.05
    if 'c2' not in params:
        params['c2'] = 2.05
    if 'con' not in params and params['wdamp'] != 1 and params['w'] != 1:
        params['con'] = 0.7298437881283576
    if 'display_info' not in params:
        params['display_info'] = 1
    if 'early_stopping' not in params:
        params['early_stopping'] = 0
    if 'early_stopping_rounds' not in params:
        params['early_stopping_rounds'] = 10
    if 'BPSO' not in params:
        params['BPSO'] = 0
    if 'velocity_limit_scale' not in params and params['BPSO'] == 0:
        params['velocity_limit_scale'] = 0.2
    elif params['BPSO'] == 1:
        params['velocity_limit_scale'] = 1

    return params


class Particle:
    def __init__(self):
        self.position = 0
        self.velocity = 0
        self.pbest_position = self.position
        self.pbest_value = float('inf')

    def initial_position(self, varmin, varmax, BPSO):
        if not BPSO:
            self.position = np.random.uniform(low=varmin, high=varmax, size=[1, len(varmin)])[0]
        else:
            self.position = np.random.uniform(low=0, high=1, size=[1, len(varmin)])[0]
        self.pbest_position = self.position
        self.velocity = np.random.uniform(low=varmin, high=varmax, size=[1, len(varmin)])[0]

    def update_velocity(self, con, w, c1, c2, gbest_position, min_velocity, max_velocity):
        self.velocity = con * ((w * self.velocity)
                               + (c1 * np.random.rand() * (self.pbest_position - self.position))
                               + (c2 * np.random.rand() * (gbest_position - self.position)))
        # Apply velocity limits
        self.velocity = np.max([self.velocity, min_velocity], axis=0)
        self.velocity = np.min([self.velocity, max_velocity], axis=0)

    def update_position(self, varmin, varmax, BPSO):
        if not BPSO:
            self.position = self.position + self.velocity
            # Apply position limits
            self.position = np.max([self.position, varmin], axis=0)
            self.position = np.min([self.position, varmax], axis=0)
        else:
            sig_velocity = self.sigmoid(self.velocity)
            self.position = np.array([1 if np.random.rand(1) < p else 0 for p in sig_velocity])

    def sigmoid(self, x):
        return 1.0/(1.0 + np.exp(-x))


def examine_particles(particles_vector, gbest_position):
    converged = 0
    ham_dist = []
    for particle in particles_vector:
        if all(particle.position == gbest_position):
            converged += 1
        ham_dist.append(sum(particle.position != gbest_position))
    return {'number_converged': converged, 'particle_hamming_distance': ham_dist}


def PSO(problem, params, *args):
    """
    Syntax: gbest_value, gbest_position, best_costs = PSO.PSO(problem, params)

    Inputs:
        problem       - This is a mandatory dictionary that contains the cost function, parameter minimum vales
                        and parameter maximum values
        params        - These are the parameters needed to specify PSO parameters. Ideally specified by the user but
                        default values are available

    Outputs:
        gbest_value     - This is the best performance achieved by PSO
        gbest_position  - This is the set of parameters that gave the best performance
        best_costs      - This is a numpy array containing the costs achieved for each iteration

    Example:
        import numpy as np
        import PSO

        print('Find the centre of a circle of the form (x-a)^2 + (y-b)^2 = r^2 where x, y and r equal 3, 4 and 5.')


        # Define the cost function
        def circle_cost(positions):
            return np.abs(25 - ((3 - positions[0]) ** 2 + (4 - positions[1]) ** 2))


        cost_function = circle_cost

        # Set maximum values of a and b
        varmin = [-1000, -1000]  # a min, b min
        varmax = [1000, 1000]  # a max, b max

        # Create problem dictionary
        problem = {'cost_function': cost_function, 'varmin': varmin, 'varmax': varmax}

        # Create PSO parameters dictionary
        params = {'maxit': 15, 'n_part': 30, 'c1': 2, 'c2': 2, 'w': 0.9, 'wdamp': 0.2, 'con': 1, 'display_info': 1,
                  'early_stopping': 1, 'early_stopping_rounds': 7}

        # Run PSO and return lowest cost, a,b values and the costs per iteration
        gbest_value, gbest_position, best_costs = PSO.PSO(problem, params)

        print('Found the centre of the circle as a = %f and b = %f' % (gbest_position[0], gbest_position[1]))
        print('The error was %f' % gbest_value)

    """
    if not args:
        args = []

    # Setting the random seed based off the current time
    np.random.seed(int(time.time()))

    cost_function = problem['cost_function']  # This tells PSO what the cost function is
    varmin = np.array(problem['varmin'])  # list of minimum lower bound values
    varmax = np.array(problem['varmax'])  # list of maximum upper bound values

    params = default_PSO_params(params)

    maxit = params['maxit']  # maximum number of PSO iterations
    n_part = params['n_part']  # number of PSO particles

    w = params['w']  # inertia coefficient
    wdamp = params['wdamp']  # damping factor for inertia coefficient
    c1 = params['c1']  # personal acceleration coefficient
    c2 = params['c2']  # global acceleration coefficient
    con = params['con']  # constriction value

    display_info = params['display_info']  # flag to show iteration info
    max_velocity = params['velocity_limit_scale'] * (varmax - varmin)
    min_velocity = -max_velocity


    particles_vector = [Particle() for _ in range(n_part)]  # create the particles


    gbest_value = float('inf')
    gbest_position = []

    for particle in particles_vector:  # set particles initial position within confines
        particle.initial_position(varmin, varmax, params['BPSO'])
        particle.pbest_position = particle.position
        particle.pbest_value = cost_function(particle.position, args)
        if particle.pbest_value < gbest_value:
            gbest_value = particle.pbest_value
            gbest_position = particle.pbest_position

    best_costs = np.zeros(maxit+1)
    best_costs[0] = gbest_value  # track best costs
    particle_debug = []
    # Enter main PSO loop
    for it in range(maxit):
        particle_debug.append(examine_particles(particles_vector, gbest_position))
        for particle in particles_vector:
            # Update velocity
            particle.update_velocity(con, w, c1, c2, gbest_position, min_velocity, max_velocity)

            # Update position
            particle.update_position(varmin, varmax, params['BPSO'])

            # Evaluate the cost at the new function
            cost_at_new_position = cost_function(particle.position, args)
            if cost_at_new_position < particle.pbest_value:
                particle.pbest_value = cost_at_new_position
                particle.pbest_position = particle.position
                if cost_at_new_position < gbest_value:
                    gbest_value = cost_at_new_position
                    gbest_position = particle.pbest_position

        best_costs[it+1] = gbest_value

        if display_info:
            print('Iteration: ', str(it), '  Best cost: ', str(gbest_value), '  Best position: ', str(gbest_position))

        w = w * wdamp

        if params['early_stopping'] and it > params['early_stopping_rounds']:
            if np.abs(best_costs[it+1] - best_costs[it+1 - params['early_stopping_rounds']]) < 0.0001:

                if display_info:
                    print('Stopping at round: ', str(it), ' as there was no performnce increase in ',
                          str(params['early_stopping_rounds']), 'rounds')
                    print('Best cost: ', str(gbest_value), '  Best position: ', str(gbest_position))
                return gbest_value, gbest_position, best_costs

    if display_info:
        print('Finished the PSO process')
        print('Best cost: ', str(gbest_value), '  Best position: ', str(gbest_position))
    return gbest_value, gbest_position, best_costs
